File: sort/quick_sort/quick_sort.cpp
/*************************************************************************
	> File Name: quick_sort.cpp
	> Author: liuyang
	> Mail: liu-yang91@qq.com 
	> Created Time: 2019年05月12日 星期日 11时43分07秒
 ************************************************************************/

#include<iostream>
using namespace std;

int quick_sort(int data[], int start, int end)
{
	if (start >= end) {
		return data[start];
	}
	else{
		int base = data[start];
		int i = start;
		int j = end; 
		while(i < j) {
			for (j; j > start; j--) {
				if (data[j] < base) {
					int tmp;
					tmp = data[i];
					data[i] = data[j];
					data[j] = tmp;
					break;
				}
				if(i>=j){
					break;
				}
			}	

			for (i; i < end; i++) {
				if (data[i] > base) {
					int tmp;
					tmp = data[i];
					data[i] = data[j];
					data[j] = tmp;
					break;
				}
				if(i>=j){
					break;
				}
			}
		}
		quick_sort(data, start, i-1);
		quick_sort(data, i+1, end);
	}
}

int main()
{
	int data[] = {10, 7, 9, 13, 4, 3, 6, 9, 0, 5, 21};
	quick_sort(data, 0, 10);
	for (int i=0; i<11; i++) {
		cout << data[i]<< " ";
	}
	cout << endl;
}

File: sort/quick_sort/quick_sort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#########################################################################
# File Name: quick_sort.py
# Author: liuyang
# mail: liu-yang91@qq.com
# Created Time: 2019年05月11日 星期六 09时37分49秒
#########################################################################

def quick_sort(data):
    if len(data) >= 2:
        mid = data[len(data)//2]
        left, right = [], []
        data.remove(mid)
        for num in data:
            if num >= mid:
                right.append(num)
            else:
                left.append(num)
        return quick_sort(left) + [mid] + quick_sort(right)
    else:
        return data

if "__main__" == __name__:
    num_list = [6,3,7,4,9,9,8,4,1]
    print(quick_sort(num_list))
